metrics/metrics.go

```diff
@@ -38,6 +40,33 @@ func NewRegistry(version string) *prometheus.Registry {
 	return registry
 }
 
+var providerMetricsKey = "provider-metrics"
+
+func ProviderCallMetricsFromContext(c context.Context) *prometheus.HistogramVec {
+	if raw := c.Value(providerMetricsKey); raw != nil {
+		return raw.(*prometheus.HistogramVec) // nolint:forcetypeassert
+	}
+	return nil
+}
+
+type ProviderActionOpts struct {
+	ProviderID      uid.ID
+	OrgID           uid.ID
+	Action          string
+	DurationSeconds float64
+}
+
+func ProviderAction(c context.Context, opts ProviderActionOpts) {
+	m := ProviderCallMetricsFromContext(c)
```

Comment:
This is overkill. Instead, this should be moved to `provideruser.go`. You're not really reducing duplication with this function